lib/Net/Async/Blockchain/BTC.pm

```diff
@@ -145,168 +88,15 @@ L<Ryu::Source>
-    die sprintf("%s: Can't reach response for block %s", $self->currency_symbol, $block_hash) unless $block_response;
-
-    my @transactions = map { $_->{block} = $block_response->{height}; $_ } $block_response->{tx}->@*;
-
-    await fmap_void { $self->transform_transaction(shift) } foreach => \@transactions,
-        concurrent                                                  => 3;
-
-    return $block_response->{height};
-}
-
-=head2 transform_transaction
-
-Receive a decoded raw transaction and convert it to a L<Net::Async::Blockchain::Transaction> object
-
-=over 4
-
-=item * C<decoded_raw_transaction> the response from the command `decoderawtransaction`
-
-=back
-
-L<Net::Async::Blockchain::Transaction>
-
-=cut
-
-async sub transform_transaction {
-    my ($self, $decoded_raw_transaction) = @_;
-
-    # this will guarantee that the transaction is from our node
-    # txindex must to be 0
-    my $received_transaction = await $self->rpc_client->get_transaction($decoded_raw_transaction->{txid});
-
-    # transaction not found, just ignore.
-    return undef unless $received_transaction;
-
-    my $fee   = Math::BigFloat->new($received_transaction->{fee} // 0);
-    my $block = Math::BigInt->new($decoded_raw_transaction->{block});
-    my @transactions;
-    my %addresses;
-
-    # we can have multiple details when:
-    # - multiple `to` addresses transactions
-    # - sent and received by the same node
-    for my $tx ($received_transaction->{details}->@*) {
-        my $address = $tx->{address};
-
-        next if $addresses{$address}++;
-
-        my @details = grep { $_->{address} eq $address } $received_transaction->{details}->@*;
-
-        my $amount = Math::BigFloat->bzero();
-        my %categories;
-        for my $detail (@details) {
-            $amount->badd($detail->{amount});
-            $categories{$detail->{category}} = 1;
-        }
-
-        my @categories       = keys %categories;
-        my $transaction_type = scalar @categories > 1 ? 'internal' : $categories[0];
-
-        my $transaction = Net::Async::Blockchain::Transaction->new(
-            currency     => $self->currency_symbol,
-            hash         => $decoded_raw_transaction->{txid},
-            block        => $block,
-            from         => '',
-            to           => $address,
-            amount       => $amount,
-            fee          => $fee,
-            fee_currency => $self->currency_symbol,
-            type         => $transaction_type,
-            timestamp    => $received_transaction->{blocktime},
-        );
-
-        push(@transactions, $transaction);
-    }
-
-    for my $transaction (@transactions) {
-        $self->source->emit($transaction);
-    }
-
-    return 1;
+    my $subscription_event = $subscription_dictionary{$subscription} or die "Invalid or not implemented subscription";
+    return $self->zmq_client->subscribe($subscription_event)->map(
```

Comment:
question : could you please help me understand the use of `map` here?

Reply:
The [Ryu source map](https://metacpan.org/pod/Ryu::Source#map) is a bit similar to the origin Perl map function [(Perl map document)](https://perldoc.perl.org/functions/map)
it performs some operations on the item then it returns the result, in our case, we formate the message then we return a formatted message.